sympy/solvers/recurr.py

```diff
@@ -822,3 +822,126 @@ def rsolve(f, y, init=None):
+    ==========
+
+    .. [1] Zeilberger, Doron. "The C-finite ansatz." The Ramanujan
+        Journal 31.1-2 (2013): 23-32.
+
+    See Also
+    ========
+    guess_seq
+
+    >>> from sympy.solvers.recurr import rfind
+    >>> from sympy.abc import x,y
+
+    >>> rfind([1,2,3,4,5,6])
+    a(n + 2) == -a(n) + 2*a(n + 1)
+    >>> rfind([0,1,1,2,3,5,8,13])
+    a(n + 2) == a(n) + a(n + 1)
+    >>> rfind([2,3,5,9,17,33,65])
+    a(n + 2) == -2*a(n) + 3*a(n + 1)
+    >>> rfind([1/3,1/9,1/27,1/81,1/243,1/729])
+    a(n + 1) == 0.333333333333333*a(n)
+
+    >>> rfind([1,y+x,y**2+2*x*y+x**2,y**3+3*x*y**2+3*x**2*y+x**3,(y+x)**4,(y+x)**5])
+    a(n + 1) == (x + y)*a(n)
+    >>> rfind([x,y,-x+2*y,-2*x+3*y,-3*x+4*y,-4*x+5*y])
+    a(n + 2) == -a(n) + 2*a(n + 1)
+
+    """
+    from sympy import Symbol, Function, Eq, simplify
+    coeffs = _rfind_helper(seq)
+    if coeffs is not None:
+        a = Function('a')
+        n = Symbol('n', integer=True)
+        cl = len(coeffs)
+        return Eq(a(n+cl-1),
+                  sum(simplify(-coeffs[i]/coeffs[cl-1]) * a(n+i)
+                      for i in range(cl-1)))
+
+
+def _rfind_helper(seq):
+    """
+    Helper function that finds linear recurrence relations
+    in a sequence of expression.
+    """
+    from sympy.core.singleton import S
+    from sympy import ceiling, Matrix
+    from numpy import trim_zeros
+
+    l = len(seq)
+    if l < 6:
+        raise ValueError('Sequence too short for guessing. Please give at least 6 values.')
+    hl = ceiling(S(l)/2)
+    A = Matrix([seq[k:k+hl] for k in range(hl)])
+    try:
+        K = A.nullspace()[0]
+    except IndexError:
+        return
+
+    return trim_zeros(K.transpose().tolist()[0])
+
+
+def guess_seq(seq):
+    """
+    Return closed form depending on `n` that produces the sequence,
+    on the condition that the first member of sequence has `n=0`.
+    Return None if no result is found.
+
+    Does only C-finite sequences currently.
+
+    >>> from sympy.solvers.recurr import guess_seq
+    >>> from sympy.abc import x,y
+
+    >>> guess_seq([1,2,3,4,5,6])
+    n + 1
+    >>> guess_seq([0,1,1,2,3,5,8,13])
+    sqrt(5)*(1/2 + sqrt(5)/2)**n/5 - sqrt(5)*(-sqrt(5)/2 + 1/2)**n/5
+    >>> guess_seq([2,3,5,9,17,33,65])
+    2**n + 1
+    >>> guess_seq([1/3,1/9,1/27,1/81,1/243,1/729])
+    0.333333333333333*0.333333333333333**n
+
+    >>> guess_seq([1*x,3*x,6*x,10*x,15*x,21*x,28*x])
+    x*(n**2/2 + 3*n/2 + 1)
+    >>> guess_seq([1,y+x,y**2+2*x*y+x**2,y**3+3*x*y**2+3*x**2*y+x**3,(y+x)**4,(y+x)**5])
+    (x + y)**n
+    >>> guess_seq([x,y,-x+2*y,-2*x+3*y,-3*x+4*y,-4*x+5*y])
+    n*(-x + y) + x
+
+    """
+    from sympy import Symbol, Function
+
+    coeffs = _rfind_helper(seq)
+    if coeffs is not None:
+        a = Function('a')
+        n = Symbol('n', integer=True)
+        cl = len(coeffs)
+        eq = sum(coeffs[i]*a(n+i) for i in range(cl))
+        start = {a(i): seq[i] for i in range(cl-1)}
+        return rsolve(eq, a(n), start)
```

Comment:
No newline at the end of file.  Please [run tests](https://github.com/sympy/sympy/wiki/Development-workflow#be-sure-that-all-tests-of-sympy-pass).